### PR TITLE
Implement MoveActorDeltaPacket

### LINKED ISSUE #5568: Make use of MoveActorDeltaPacket
## Description
This has been in the protocol for ages and offers a fairly easy bandwidth win by reducing outflow of non-cacheable, poorly-compressible data (entity position updates).

This wasn't implemented initially due to technical complications with the packet design, but those have long since been resolved, and it should be pretty easy to implement this now.

### PR DESCRIPTION
**Pull request by AkmalFairuz**
## Introduction
Implement MoveActorDeltaPacket

### Relevant issues
Fix https://github.com/pmmp/PocketMine-MP/issues/5568

### CODE REVIEW

File: src/entity/Entity.php [hunk 1]
```diff
@@ -785,18 +789,50 @@ protected function broadcastMovement(bool $teleport = false) : void{
...
+				if(($this->location->yaw - $this->lastBroadcastedMovement->yaw) !== 0.0){
+					$pk->yaw = $this->location->yaw;
+					$pk->flags |= MoveActorDeltaPacket::FLAG_HAS_YAW;
+					$pk->headYaw = $this->location->yaw;
+					$pk->flags |= MoveActorDeltaPacket::FLAG_HAS_HEAD_YAW;
+				}
+				$this->server->broadcastPackets($this->hasSpawned, [$pk]);
+			}
 		}
+		$this->lastBroadcastedMovement = Location::fromObject($offsetPosition, null, $this->location->yaw, $this->location->pitch);
```
Comment (dktapps) [inline src/entity/Entity.php]: We already have `lastLocation` for this.

**Review by dktapps — CHANGES_REQUESTED** [review]
The position used in sendSpawnPacket needs to be changed to use the last broadcast position as well, or the initial delta may accumulate errors.

File: src/entity/Entity.php [hunk 2]
```diff
@@ -1477,7 +1495,7 @@ protected function sendSpawnPacket(Player $player) : void{
 			$this->getId(), //TODO: actor unique ID
 			$this->getId(),
 			static::getNetworkTypeId(),
-			$this->location->asVector3(),
+			$this->lastLocation->asVector3(),
```
Comment (dktapps) [inline src/entity/Entity.php]: This needs to be changed in subclasses as well.

**Comment by dktapps** [comment]
It occurs to me that it may not be possible to ensure that plugin devs use the correct position in custom sendSpawnPacket implementations. If an incorrect position is used, the delta position will accumulate errors and the source of the problem will be far from obvious.

Any suggestions how to combat this?

File: src/entity/Entity.php [hunk 3]
```diff
@@ -785,17 +786,34 @@ protected function broadcastMovement(bool $teleport = false) : void{
...
-				$this->getOffsetPosition($this->location),
-				$this->location->pitch,
-				$this->location->yaw,
-				$this->location->yaw,
-				(
-					//TODO: if the above hack for #4394 gets removed, we should be setting FLAG_TELEPORT here
-					($this->onGround ? MoveActorAbsolutePacket::FLAG_GROUND : 0)
-				)
-			)]);
+			$deltaLoc = $offsetPosition->subtractVector($this->getOffsetPosition($this->lastLocation));
```
Comment (dktapps) [inline src/entity/Entity.php]: This calculation is useless. With the current code, you can directly compare lastLocation->whatever with currentLocation->whatever.

File: src/entity/Entity.php [hunk 4]
```diff
@@ -785,17 +786,34 @@ protected function broadcastMovement(bool $teleport = false) : void{
...
+			}
+			if($deltaLoc->y != 0.0){
+				$pk->yPos = $offsetPosition->y;
+				$pk->flags |= MoveActorDeltaPacket::FLAG_HAS_Y;
+			}
+			if($deltaLoc->z !== 0.0){
+				$pk->zPos = $offsetPosition->z;
+				$pk->flags |= MoveActorDeltaPacket::FLAG_HAS_Z;
+			}
+			if(($this->location->pitch - $this->lastLocation->pitch) !== 0.0){
```
Comment (dktapps) [inline src/entity/Entity.php]: This subtraction is pointless. It's not telling you anything that an equality operator wouldn't.

File: src/entity/Entity.php [hunk 5]
```diff
@@ -785,17 +786,34 @@ protected function broadcastMovement(bool $teleport = false) : void{
...
+				$pk->pitch = $this->location->pitch;
+				$pk->flags |= MoveActorDeltaPacket::FLAG_HAS_PITCH;
+			}
+			if(($this->location->yaw - $this->lastLocation->yaw) !== 0.0){
```
Comment (dktapps) [inline src/entity/Entity.php]: Same here.

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/entity/Entity.php [hunk 6]
```diff
@@ -1477,7 +1495,7 @@ protected function sendSpawnPacket(Player $player) : void{
 			$this->getId(), //TODO: actor unique ID
 			$this->getId(),
 			static::getNetworkTypeId(),
-			$this->location->asVector3(),
+			$this->lastLocation->asVector3(),
 			$this->getMotion(),
 			$this->location->pitch,
 			$this->location->yaw,
```
Comment (dktapps) [inline src/entity/Entity.php]: These also need to use `lastLocation`.

**Comment by AkmalFairuz** [comment]
> It occurs to me that it may not be possible to ensure that plugin devs use the correct position in custom sendSpawnPacket implementations. If an incorrect position is used, the delta position will accumulate errors and the source of the problem will be far from obvious.
> 
> Any suggestions how to combat this?

Store players who need an absolute move packet when spawned to players, then send absolute move packet when movement packet is broadcasted. See my commit

**Comment by dktapps** [comment]
You could probably reduce it to a boolean flag. Not much sense in adding the extra memory usage and complexity. It anyway might be useful to send periodic absolute updates anyway.

**Comment by dktapps** [comment]
I meant a single boolean, not an array of booleans.

File: src/entity/Entity.php [hunk 7]
```diff
@@ -220,6 +221,9 @@ public static function nextRuntimeId() : int{
 	/** @var int|null */
 	protected $targetId = null;
 
+	/** @var bool */
+	protected $needsAbsoluteMovement = true;
```
Comment (dktapps) [inline src/entity/Entity.php:225]: Use native types for this. Also, this should be `private`.

File: src/entity/Entity.php [hunk 8]
```diff
@@ -785,17 +790,48 @@ protected function broadcastMovement(bool $teleport = false) : void{
 				$this->spawnTo($player);
```
Comment (dktapps) [inline src/entity/Entity.php:790]: This is no longer going to work correctly, since it's now using the lastLocation.

**Pull request closed without merging**